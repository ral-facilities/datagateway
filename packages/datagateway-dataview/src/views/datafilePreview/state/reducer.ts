import DATAFILE_PREVIEWER_DEFAULT from '../defaults';
import {
  DecrementDatafilePreviewerZoomLevelType,
  IncrementDatafilePreviewerZoomLevelType,
  ResetDatafilePreviewerZoomLevelType,
  ToggleDatafilePreviewerDetailsPanePayload,
  ToggleDatafilePreviewerDetailsPaneType,
} from './actions';
import { DGDataViewState } from '../../../state/app.types';

/**
 * Defines the shape of the redux state for {@link DatafilePreviewer}
 */
interface DatafilePreviewerState {
  /**
   * Whether the details pane should be shown or not.
   */
  isDetailsPaneShown: boolean;

  /**
   * The current zoom level of the previewer, in percentage. For example,
   * 100% maps to 100, 110% maps to 110, 80% maps to 80, etc.
   */
  zoomLevel: number;
}

function toggleDatafilePreviewerDetailsPane(
  state: DGDataViewState,
  payload: ToggleDatafilePreviewerDetailsPanePayload
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      isDetailsPaneShown: payload.shouldShow,
    },
  };
}

function incDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      zoomLevel:
        state.datafilePreviewer.zoomLevel +
        DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep,
    },
  };
}

function decDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      // zoom level should not go lower than zoomLevelStep
      // otherwise will result in zoom level <= 0.
      zoomLevel: Math.max(
        DATAFILE_PREVIEWER_DEFAULT.minZoomLevel,
        state.datafilePreviewer.zoomLevel -
          DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep
      ),
    },
  };
}

function resetDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
    },
  };
}

const datafilePreviewerInitialState: DatafilePreviewerState = {
  isDetailsPaneShown: DATAFILE_PREVIEWER_DEFAULT.showDetailsPane,
  zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
};

const datafilePreviewerReducer = {
  [ToggleDatafilePreviewerDetailsPaneType]: toggleDatafilePreviewerDetailsPane,
  [IncrementDatafilePreviewerZoomLevelType]: incDatafilePreviewerZoomLevel,
  [DecrementDatafilePreviewerZoomLevelType]: decDatafilePreviewerZoomLevel,
  [ResetDatafilePreviewerZoomLevelType]: resetDatafilePreviewerZoomLevel,
};

export {
  datafilePreviewerInitialState,
  datafilePreviewerReducer,
  toggleDatafilePreviewerDetailsPane,
  incDatafilePreviewerZoomLevel,
  decDatafilePreviewerZoomLevel,
  resetDatafilePreviewerZoomLevel,
};
export type { DatafilePreviewerState };
