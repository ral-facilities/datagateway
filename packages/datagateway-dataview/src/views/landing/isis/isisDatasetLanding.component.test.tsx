import React from 'react';
import ISISDatasetLanding from './isisDatasetLanding.component';
import { initialState as dgDataViewInitialState } from '../../../state/reducers/dgdataview.reducer';
import configureStore from 'redux-mock-store';
import { StateType } from '../../../state/app.types';
import {
  Dataset,
  dGCommonInitialState,
  useDatasetDetails,
  useDatasetSizes,
} from 'datagateway-common';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { Typography } from '@mui/material';
import { mount, ReactWrapper } from 'enzyme';
import { createMemoryHistory, History } from 'history';
import { QueryClientProvider, QueryClient } from 'react-query';
import { Router } from 'react-router-dom';

jest.mock('datagateway-common', () => {
  const originalModule = jest.requireActual('datagateway-common');

  return {
    __esModule: true,
    ...originalModule,
    useDatasetDetails: jest.fn(),
    useDatasetSizes: jest.fn(),
  };
});

describe('ISIS Dataset Landing page', () => {
  const mockStore = configureStore([thunk]);
  let state: StateType;
  let history: History;

  const createWrapper = (studyHierarchy = false): ReactWrapper => {
    const store = mockStore(state);
    return mount(
      <Provider store={store}>
        <Router history={history}>
          <QueryClientProvider client={new QueryClient()}>
            <ISISDatasetLanding
              instrumentId="4"
              instrumentChildId="5"
              investigationId="1"
              datasetId="87"
              studyHierarchy={studyHierarchy}
            />
          </QueryClientProvider>
        </Router>
      </Provider>
    );
  };

  const initialData: Dataset = {
    id: 87,
    name: 'Test 1',
    description: 'foo bar',
    modTime: '2019-06-10',
    createTime: '2019-06-10',
    doi: 'doi 1',
    startDate: '2019-06-10',
    endDate: '2019-06-11',
    complete: true,
    type: {
      id: 1,
      name: 'Type 1',
      description: 'The first type',
    },
  };
  beforeEach(() => {
    state = JSON.parse(
      JSON.stringify({
        dgdataview: dgDataViewInitialState,
        dgcommon: dGCommonInitialState,
      })
    );
    history = createMemoryHistory();

    (useDatasetDetails as jest.Mock).mockReturnValue({
      data: initialData,
    });
    (useDatasetSizes as jest.Mock).mockReturnValue({
      data: 1,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('calls the correct data fetching hooks', () => {
    createWrapper();

    expect(useDatasetDetails).toHaveBeenCalledWith(87);
    expect(useDatasetSizes).toHaveBeenCalledWith(initialData);
  });

  it('links to the correct url in the datafiles tab for both hierarchies and both views', () => {
    const facilityCycleWrapper = createWrapper();

    facilityCycleWrapper
      .find('#dataset-datafiles-tab')
      .last()
      .simulate('click');

    expect(history.location.pathname).toBe(
      '/browse/instrument/4/facilityCycle/5/investigation/1/dataset/87/datafile'
    );

    history.replace('/?view=card');
    const studyWrapper = createWrapper(true);

    studyWrapper.find('#dataset-datafiles-tab').last().simulate('click');

    expect(history.location.pathname).toBe(
      '/browseStudyHierarchy/instrument/4/study/5/investigation/1/dataset/87/datafile'
    );
    expect(history.location.search).toBe('?view=card');
  });

  it('displays DOI and renders the expected Link ', () => {
    const wrapper = createWrapper();

    expect(
      wrapper
        .find('[data-testid="isis-dataset-landing-doi-link"]')
        .first()
        .text()
    ).toEqual('doi 1');

    expect(
      wrapper
        .find('[data-testid="isis-dataset-landing-doi-link"]')
        .first()
        .prop('href')
    ).toEqual('https://doi.org/doi 1');
  });

  it('useDatasetSizes queries not sent if no data returned from useDatasetDetails', () => {
    (useDatasetDetails as jest.Mock).mockReturnValue({
      data: undefined,
    });
    createWrapper();
    expect(useDatasetSizes).toHaveBeenCalledWith(undefined);
  });

  it('incomplete datasets render correctly', () => {
    initialData.complete = false;
    (useDatasetDetails as jest.Mock).mockReturnValue({
      data: initialData,
    });
    const wrapper = createWrapper();

    expect(wrapper.find(Typography).last().text()).toEqual(
      'datasets.incomplete'
    );
  });
});
