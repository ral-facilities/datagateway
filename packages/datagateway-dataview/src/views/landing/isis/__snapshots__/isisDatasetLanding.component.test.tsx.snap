// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ISIS Dataset Landing page renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <LandingPage
    data={
      Array [
        Object {
          "complete": true,
          "createTime": "2019-06-10",
          "description": "foo bar",
          "doi": "doi 1",
          "endDate": "2019-06-11",
          "id": 87,
          "modTime": "2019-06-10",
          "name": "Test 1",
          "size": 1,
          "startDate": "2019-06-10",
        },
      ]
    }
    datasetId="87"
    fetchData={[Function]}
    fetchDetails={[Function]}
    instrumentChildId="5"
    instrumentId="4"
    investigationId="1"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    studyHierarchy={false}
    view={null}
    viewDatafiles={[Function]}
  />
</ContextProvider>
`;
